### PR TITLE
mCODE Data + Report Modal

### PR DESCRIPTION
**Pull request by blangley28**
Creates a new CQL file to grab mCODE data from the patient record and then make it available in the patient provider. Then it uses this data to populate the report modal.

Screen Shot 2020-04-23 at 2 35 14 PM

Putting this PR up now for comments but Keeyan's branch should be merged into this first.
Note: eslint fails but this is fixed by Keeyan's PR

### CODE REVIEW

File: public/static/cql/mCODE.cql [hunk 1]
```diff
@@ -4,9 +4,121 @@ using FHIR version '4.0.0'
...
 codesystem "SNOMEDCT": 'http://snomed.info/sct'
+codesystem "LOINC": 'http://loinc.org'
+codesystem "OBSCAT": 'http://terminology.hl7.org/CodeSystem/observation-category'
+
+// FHIR CODES
+code "Primary cancer code": '64572001' from "SNOMEDCT" display 'Disease (disorder)'
+code "TNM clinical primary tumor code": '21905-5' from "LOINC" display 'Primary tumor.clinical [Class] Cancer'
+code "TNM clinical regional nodes code": '21906-3' from "LOINC" display 'Regional lymph nodes.clinical [Class] Cancer'
+code "TNM clinical distant metastases code": '21907-1' from "LOINC" display 'Distant metastases.clinical [Class] Cancer'
+code "Tumor marker test code": 'laboratory' from "OBSCAT"
```
Comment (radamson) [inline public/static/cql/mCODE.cql]: Should we have the display here like with the other codes?

File: public/static/cql/mCODE.cql [hunk 2]
```diff
@@ -4,9 +4,121 @@ using FHIR version '4.0.0'
...
+code "Estrogen receptor code": '85337-4' from "LOINC" display 'Estrogen receptor Ag [Presence] in Breast cancer specimen by Immune stain'
+code "Progesterone receptor code": '85339-0' from "LOINC" display 'Progesterone receptor Ag [Presence] in Breast cancer specimen by Immune stain'
+code "HER2 receptor code": '85319-2' from "LOINC" display 'HER2 [Presence] in Breast cancer specimen by Immune stain'
+
+// selected cancer codes used within Synthea
+// not the entire list supported by PrimaryCancerConditionVS
+code "11p partial monosomy syndrome 4135001 code": '4135001' from SNOMEDCT display '11p partial monosomy syndrome'
```
Comment (radamson) [inline public/static/cql/mCODE.cql]: Is there a reason these are declared with the code in the name?

i.e.

`11p partial monosomy syndrome 4135001 code`

vs

`11p partial monosomy syndrome`

File: src/components/App.tsx [hunk 3]
```diff
@@ -42,14 +47,61 @@ const App: FC<AppProps> = ({ demoId }) => {
...
+      .then(elm => {
+        let elmResults: ElmResults = {
+          patientResults: {}
+        };
+        elmResults = executeElm(bundle, elm);
+        const patientIds = Object.keys(elmResults.patientResults);
+        const mcodeData = elmResults.patientResults[patientIds[0]];
+
+        const mcodeElements: McodeElements = {
+          primaryCancer: mcodeData['Primary Cancer Condition Code'][0] ?? undefined,
```
Comment (radamson) [inline src/components/App.tsx]: feels weird to be setting something to undefined while defining it.  Does null work here?

File: src/components/App.tsx [hunk 4]
```diff
@@ -42,14 +47,61 @@ const App: FC<AppProps> = ({ demoId }) => {
...
+    // Create a Bundle for the CQL engine
+    const bundle = {
+      resourceType: 'Bundle',
+      type: 'searchset',
+      entry: resources.map((r: DomainResource) => ({ resource: r }))
+    };
+    getFixture('mCODE.cql')
+      .then(cql => convertBasicCQL(cql))
+      .then(elm => {
+        let elmResults: ElmResults = {
```
Comment (radamson) [inline src/components/App.tsx]: This variable initializer is redundant.  You can remove it and just use the definition on line 63.

File: src/components/App.tsx [hunk 5]
```diff
@@ -42,14 +47,61 @@ const App: FC<AppProps> = ({ demoId }) => {
...
+          patientResults: {}
+        };
+        elmResults = executeElm(bundle, elm);
+        const patientIds = Object.keys(elmResults.patientResults);
```
Comment (radamson) [inline src/components/App.tsx]: I suppose any resources could be passed into this function, but you should also be able to just get the id out of the `patient.id` state.  I'm not sure if there is an advantage to doing it one way or another, but at first glance it seems advantageous to grab it from the already available source.

File: src/utils/fhirUtils.ts [hunk 6]
```diff
@@ -211,3 +212,13 @@ export function createCarePlan(title: string, patient: Patient): CarePlan {
     subject: { reference: `Patient/${patient.id}` }
   };
 }
+
+export function getTNM(mcodeElements: McodeElements): string {
+  let tnm = '';
```
Comment (radamson) [inline src/utils/fhirUtils.ts]: clearly this works, but its a lot of `+=` and `??` to get the whitespace right.  Could probably use join here (and avoid a trailing whitespace if there is no `metastatsesCategory`)

I didn't actually test this so I might have a syntax error.

```typescript
return [ mcodeElements.tumorCategory?.split(' ')[0],
              mcodeElements.nodeCategory?.split(' ')[0],
              mcodeElements.metastasesCategory?.split(' ')[0]
           ].join(' ')

File: src/components/ReportModal/ReportModal.tsx [hunk 7]
```diff
@@ -21,12 +24,12 @@ const ReportModal: FC<ReportModalInterface> = ({ onConfirm, onDecline }) => {
...
   const observationField = [
-    { name: 'Primary Cancer', value: undefined },
-    { name: 'Laterality', value: undefined },
-    { name: 'Clinical TNM', value: undefined },
-    { name: 'Estrogen Receptor', value: undefined },
-    { name: 'Progesterone Receptor', value: undefined },
-    { name: 'HER2 Receptor', value: undefined }
+    { name: 'Primary Cancer', value: mcodeRecords.primaryCancer ?? 'Unknown' },
+    { name: 'Laterality', value: mcodeRecords.laterality ?? 'Unknown' },
+    { name: 'Clinical TNM', value: getTNM(mcodeRecords) },
```
Comment (radamson) [inline src/components/ReportModal/ReportModal.tsx:29]: should this one also have `?? 'Unknown'` ?

**Review by radamson — COMMENTED** [review]
Tested it out and looks good.  I didn't check for linting in this PR.  I had a couple of suggestions and questions to help confirm my understanding.

I'm always learning something when CQL is involved 🙂

File: public/static/cql/mCODE.cql (hunk 2 above)
Comment (blangley28) [reply public/static/cql/mCODE.cql]: @dehall I think you wrote these originally for the criteria? All of the `code`s include "code" at the end but as for including the numerical code I'm not sure why

File: src/components/App.tsx (hunk 3 above)
Comment (blangley28) [reply src/components/App.tsx]: I went with undefined because the types are declared as optional. Making them null is a possibility with updating the type

File: src/components/App.tsx (hunk 4 above)
Comment (blangley28) [reply src/components/App.tsx]: Let me give it a test but IIRC without explicitly defining the `patientResults` element inside `elmResults` I was unable to read the data correctly. I copied this from somewhere else in the engine and also thought it was peculiar but deleting it failed.

File: src/components/App.tsx (hunk 3 above)
Comment (radamson) [reply src/components/App.tsx]: eh, we can just leave it undefined then I forgot about changing the type.

File: src/components/App.tsx (hunk 4 above)
Comment (radamson) [reply src/components/App.tsx]: my IDE was picking it up  ¯\_(ツ)_/¯

File: src/utils/fhirUtils.ts (hunk 6 above)
Comment (blangley28) [reply src/utils/fhirUtils.ts]: It would avoid the trailing whitespace but also introduce a double space if `nodeCategory` is undefined. At the end of the day neither are a big deal and I think this suggestion is cleaner so I'll go with it

File: src/components/ReportModal/ReportModal.tsx (hunk 7 above)
Comment (blangley28) [reply src/components/ReportModal/ReportModal.tsx:29]: No, `getTNM` always returns a string. However if all three TNM values are undefined the result will be "   " so I will add a check to the method so it returns "Unknown" if all 3 are undefined

File: public/static/cql/mCODE.cql (hunk 2 above)
Comment (dehall) [reply public/static/cql/mCODE.cql]: I copied them into the mCODE_Library.cql file from an old version of the CQL on our internal repo: https://gitlab.mitre.org/mcode/pathways/-/blob/master/src/model/valuesets/PrimaryCancerVS.cql

That said I'm fine with whatever format we want to use for the code names, because we should only be referencing them internally in the definition of a concept.

File: src/components/App.tsx (hunk 5 above)
Comment (blangley28) [reply src/components/App.tsx]: Just gave this a try but I think it will cause more headache than gain since `patient` can be null and `id` can be undefined

File: public/static/cql/mCODE.cql (hunk 1 above)
Comment (blangley28) [reply public/static/cql/mCODE.cql]: Removed since this code isn't actually used

File: public/static/cql/mCODE.cql (hunk 2 above)
Comment (blangley28) [reply public/static/cql/mCODE.cql]: I removed the numerical code values for consistency with the rest of the file.
Ie: "11p partial monosomy syndrome 4135001 code" --> "11p partial monosomy syndrome code"

File: public/static/cql/mCODE.cql [hunk 8]
```diff
@@ -4,9 +4,120 @@ using FHIR version '4.0.0'
 
 // CODESYSTEMS
 codesystem "SNOMEDCT": 'http://snomed.info/sct'
+codesystem "LOINC": 'http://loinc.org'
+codesystem "OBSCAT": 'http://terminology.hl7.org/CodeSystem/observation-category'
```
Comment (radamson) [inline public/static/cql/mCODE.cql]: It doesn't look like this has any uses and can be removed now that the code using it has been removed.

Comment (blangley28) [reply public/static/cql/mCODE.cql]: Good catch. Removed

File: src/components/App.tsx [hunk 9]
```diff
@@ -42,14 +47,58 @@ const App: FC<AppProps> = ({ demoId }) => {
...
 
+  const setMcodeRecords = useCallback((resources: DomainResource[]): void => {
+    // Create a Bundle for the CQL engine
+    const bundle = {
+      resourceType: 'Bundle',
+      type: 'searchset',
+      entry: resources.map((r: DomainResource) => ({ resource: r }))
+    };
+    getFixture('mCODE.cql')
+      .then(cql => convertBasicCQL(cql))
```
Comment (dehall) [inline src/components/App.tsx]: Instead of converting the CQL to ELM on every app load can we pre-convert it and store the ELM somewhere? Additionally maybe we should look into extracting the mCODE CQL library into its own repository again, though that's not part of this task

File: src/components/App.tsx [hunk 10]
```diff
@@ -42,14 +47,58 @@ const App: FC<AppProps> = ({ demoId }) => {
...
+      .then(elm => {
+        let elmResults = executeElm(bundle, elm);
+        const patientIds = Object.keys(elmResults.patientResults);
+        const mcodeData = elmResults.patientResults[patientIds[0]];
+
+        const mcodeElements: McodeElements = {
+          primaryCancer: mcodeData['Primary Cancer Condition Code'][0] ?? undefined,
+          laterality: mcodeData['Primary Cancer Condition Body Location Code'][0] ?? undefined,
+          tumorCategory:
+            mcodeData['TNM Clinical Primary Tumor Category Data Value (T Category)'][0] ??
```
Comment (dehall) [inline src/components/App.tsx]: To make this a little more robust, instead of just fetching the `text` field in the referenced CQL could these also look at, for example, the `code.display` field? We don't have to account for all the many possibilities here, like interpreting codes if both text and code.display are empty (which is totally legal), but test data may not always include the text field. Maybe it's a helper function in the CQL or in the code here but it shouldn't be too complex

File: src/components/App.tsx [hunk 11]
```diff
@@ -42,14 +47,58 @@ const App: FC<AppProps> = ({ demoId }) => {
     setEvaluatePath(true);
   }, []);
 
+  const setMcodeRecords = useCallback((resources: DomainResource[]): void => {
+    // Create a Bundle for the CQL engine
```
Comment (dehall) [inline src/components/App.tsx:50]: We do this in so many places I wonder if it makes more sense to just make the ELM executor take in an array instead of a bundle
Edit: this would be more appropriate as a very quick 1-pt task instead of piling on unrelated changes here

**Review by dehall — COMMENTED** [review]
Looks good -- main thing to me is preconverting the CQL to ELM

File: src/components/App.tsx (hunk 10 above)
Comment (radamson) [reply src/components/App.tsx]: That's a good point @dehall.  From my experience, and without any evidence to back it up, display is more common (and doesn't surprise me because it is taken right from the system).

Comment (blangley28) [reply src/components/App.tsx]: I'll update it to read `code[0].coding[0].display.value`

File: src/components/App.tsx (hunk 9 above)
Comment (blangley28) [reply src/components/App.tsx]: Added pre-converted elm. I agree - the task already exists in the backlog: https://jira.mitre.org/browse/PATHWAYS-117

File: src/components/App.tsx (hunk 11 above)
Comment (blangley28) [reply src/components/App.tsx:50]: https://jira.mitre.org/browse/PATHWAYS-242

**Review by radamson — APPROVED** [review]

**Review by dehall — APPROVED** [review]
